## 데이터 불러오기

### kaggle에 올라와 있는 bike-sharing-demand의 데이터 이용

https://www.kaggle.com/c/bike-sharing-demand

In [1]:
import pandas as pd

train = pd.read_csv("./data/bike-sharing-demand/train.csv")
test = pd.read_csv("./data/bike-sharing-demand/test.csv")
train=train.drop(["casual","datetime","registered"],axis=1)
print(train.shape)
print(test.shape)

(10886, 9)
(6493, 9)


In [2]:
print(train.info())     # 결측치나 데이터 개수 또는 자료형 확인
print(train.columns)    # 열 이름 확인
print(train.shape)      # 행 열 확인 
print(train.describe()) # 요약값 확인
print(train.head())     # 앞의 데이터 몇줄
print(train.tail())     # 뒤의 데이터 몇줄

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 9 columns):
season        10886 non-null int64
holiday       10886 non-null int64
workingday    10886 non-null int64
weather       10886 non-null int64
temp          10886 non-null float64
atemp         10886 non-null float64
humidity      10886 non-null int64
windspeed     10886 non-null float64
count         10886 non-null int64
dtypes: float64(3), int64(6)
memory usage: 765.5 KB
None
Index(['season', 'holiday', 'workingday', 'weather', 'temp', 'atemp',
       'humidity', 'windspeed', 'count'],
      dtype='object')
(10886, 9)
             season       holiday    workingday       weather         temp  \
count  10886.000000  10886.000000  10886.000000  10886.000000  10886.00000   
mean       2.506614      0.028569      0.680875      1.418427     20.23086   
std        1.116174      0.166599      0.466159      0.633839      7.79159   
min        1.000000      0.000000      0.000000      1.00

In [3]:
x = train.loc[:,"season":"windspeed"]
print(x.info())  
y = train['count']

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 8 columns):
season        10886 non-null int64
holiday       10886 non-null int64
workingday    10886 non-null int64
weather       10886 non-null int64
temp          10886 non-null float64
atemp         10886 non-null float64
humidity      10886 non-null int64
windspeed     10886 non-null float64
dtypes: float64(3), int64(5)
memory usage: 680.5 KB
None


## 데이터셋 나누기

In [4]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, 
                                                    test_size=0.2,
                                                    random_state=42)

## 모델 만들어 학습 및 예측

#### 1. 모델 만들기
#### 2. 모델 학습  model.fit()
#### 3. 모델 예측 model.predict()
#### 4. 모델 평가

In [5]:
from sklearn.linear_model import LinearRegression
model=LinearRegression().fit(x_train,y_train)
pred=model.predict(x_test)
pred

array([315.86982513,  40.92678568, 178.49750562, ..., 265.43804981,
        25.07105754, 202.25843959])

In [6]:
import pandas as pd
df=pd.DataFrame({"y_test" : y_test,"prediction" : pred, "diff" : (y_test - pred)})
df

y_test  prediction        diff
3133      127  315.869825 -188.869825
5786       13   40.926786  -27.926786
5224      163  178.497506  -15.497506
8953      233  280.271742  -47.271742
8054      222  258.464323  -36.464323
10044     166  200.174624  -34.174624
5337      144  235.583134  -91.583134
2753      376  350.012623   25.987377
10127     601  157.443498  443.556502
33         53   58.865862   -5.865862
4634        6   67.618217  -61.618217
8104       14  148.366862 -134.366862
2855      209  274.859513  -65.859513
3460       62  169.139791 -107.139791
1408      139  328.283480 -189.283480
3844        9  165.421274 -156.421274
9257       96  174.048509  -78.048509
1056       98  237.794107 -139.794107
3837      230  189.757141   40.242859
3079       12  228.085292 -216.085292
1608       54   65.952859  -11.952859
1768      107  191.727080  -84.727080
10720       8  168.142753 -160.142753
2507      182  230.292088  -48.292088
6906       11  174.622959 -163.622959
7051        1  116.113689 -115.113689
7361       93  138.205213  -45.205213
3487       17  154.232887 -137.232887
483        39  102.000119  -63.000119
3704       11  209.114920 -198.114920
...       ...         ...         ...
7443      280  250.108886   29.891114
10810     122  127.722331   -5.722331
2452       13  213.725639 -200.725639
9229      299  234.307198   64.692802
7923      502  355.137889  146.862111
837       272   95.313039  176.686961
7482       13  238.155885 -225.155885
5452       16  153.493579 -137.493579
3784      156  152.643350    3.356650
9021       27  169.827231 -142.827231
6109      262  118.330385  143.669615
7793        8  172.130833 -164.130833
6220      120  116.062469    3.937531
6673      623   91.307798  531.692202
10509     502  122.237019  379.762981
157        51  100.315408  -49.315408
6930       36  236.920307 -200.920307
6821      281  254.100505   26.899495
1441       46  161.937393 -115.937393
8904      305  319.471177  -14.471177
10237     109  176.770778  -67.770778
8698      397  346.001310   50.998690
5479       60   90.626564  -30.626564
4638       90   84.282999    5.717001
8241      430  385.058831   44.941169
1963      351  258.247836   92.752164
5807        2   99.377620  -97.377620
10061     509  265.438050  243.561950
5427        5   25.071058  -20.071058
7588      239  202.258440   36.741560

[2178 rows x 3 columns]

In [7]:
#MAE, MSE 가 낮을 수록 좋다.

MAE=sum(abs(df["diff"]))/len(df["diff"])
MSE=sum((df["diff"])**2)/len(df["diff"])

import numpy as np

RMSE=np.sqrt(MSE)

print(MAE)
print(MSE)
print(RMSE)

115.04130180057099
23906.814693176108
154.61828705937765


In [8]:
acc_tr=model.score(x_train,y_train)
acc_test=model.score(x_test,y_test)
print("훈련 세트의 정확도 : {:.2f}".format(acc_tr))  
print("테스트 세트의 정확도 : {:.2f}".format(acc_test))  

훈련 세트의 정확도 : 0.26
테스트 세트의 정확도 : 0.28


test_size=0.4
MAE=116.56986341453377
MSE=24043.936838228954
RMSE=155.06107454235234
훈련 세트의 정확도 : 0.26
테스트 세트의 정확도 : 0.26

test_size=0.3
MAE=115.43870449385717
MSE=24002.84808256994
RMSE=154.9285257225729
훈련 세트의 정확도 : 0.26
테스트 세트의 정확도 : 0.26

test_size=0.2
MAE=115.04130180057105
MSE=23906.814693176093
RMSE=154.6182870593776
훈련 세트의 정확도 : 0.26
테스트 세트의 정확도 : 0.28

test_size=0.25
MAE=114.85576302051685
MSE=23969.963158339433
RMSE=154.8223600076534
훈련 세트의 정확도 : 0.26
테스트 세트의 정확도 : 0.27

test_size=0.1
MAE=115.04264797974525
MSE=24106.80842551748
RMSE=155.2636738761436
훈련 세트의 정확도 : 0.26
테스트 세트의 정확도 : 0.27

### 정확도를 보았을 때 훈련용과 테스트용의 차이가 가장 적은 것 중
### MAE or RMSE가 가장 작은 test_size가 0.3인 것이 더 좋은 데이터 분류 기준